### PR TITLE
Retry callback into job queue / fix simulation_context

### PR DESCRIPTION
**Pull request by xjules**
- fix sumbit test cound and update status
- test queue retries
- test for queue manager and fix to job queue test
- fix for simulation_context and simulation_batch and the tests which use python job now
- integrates and updates #757 

**Issue**
#686 

**Approach**
It includes changing the job status and making sure number of submits is properly set.

### CODE REVIEW

**Comment by markusdregi** [comment]
@xjules The tests are failing... Are you sure this one is ready for review?

**Comment by xjules** [comment]
> @xjules The tests are failing... Are you sure this one is ready for review?

They have been working actually (for some time at least), then they started failing due waiting time on some jobs (I set the default retry after failing job to 1, which is also stated in documentation). anyway need to fix the `simulation_context` and the corresponding tests, which are still linked to previous C job queue version. All in all this would require a new issue to clean the C job queue / manager code.

File: python/res/job_queue/queue.py [hunk 1]
```diff
@@ -273,6 +278,12 @@ def fetch_next_waiting(self):
                 return job
         return None
 
+
+    def count_status(self, status):
+        return len([job for job in self.job_list if job.status == status])
+
+
```
Comment (ManInFez) [inline python/res/job_queue/queue.py]: Lots and lots of whitespace..

File: python/res/simulator/simulation_context.py [hunk 2]
```diff
@@ -19,43 +15,34 @@ def __init__(self, ert, sim_fs, mask, itr , verbose=False):
...
-        self._ert.initRun(self._run_context)
-
-
-    def __len__(self):
-        return self._mask.count()
 
+        self._run_context = ErtRunContext( EnkfRunType.ENSEMBLE_EXPERIMENT, sim_fs, None, mask, path_fmt, jobname_fmt, subst_list, itr)
+        for run_arg in self._run_context:
+            if run_arg is not None:
+                self._run_args[run_arg.iens] = run_arg
```
Comment (ManInFez) [inline python/res/simulator/simulation_context.py]: Why are we storing both the run_context and run_args. Ownership of the run_args objects is not very clearly seen here.

File: python/tests/res/simulator/test_batch_sim.py [hunk 3]
```diff
@@ -193,10 +192,8 @@ def test_batch_simulation(self):
             ]
 
             ctx = rsim.start("case", case_data)
+
```
Comment (ManInFez) [inline python/tests/res/simulator/test_batch_sim.py]: remove whitespace

File: python/tests/res/simulator/test_simulation_context.py [hunk 4]
```diff
@@ -37,31 +37,20 @@ def test_simulation_context(self):
...
-            with self.assertRaises(UserWarning):
-                simulation_context1.addSimulation(0, geo_id)
-
             while simulation_context1.isRunning():
                 time.sleep(1.0)
 
             while simulation_context2.isRunning():
                 time.sleep(1.0)
 
+
```
Comment (ManInFez) [inline python/tests/res/simulator/test_simulation_context.py]: remove whitespace

**Comment by ManInFez** [comment]
Just some minor comments. Put back in "In progress"

File: python/res/job_queue/job_queue_manager.py [hunk 5]
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
...
-
-        return ts.datetime()
-
-
     def get_job_queue(self):
         return self.queue
 
     def stop_queue(self):
-        self._stop_queue( )
+        self.get_job_queue().kill_all_jobs()
```
Comment (andreabrambilla) [inline python/res/job_queue/job_queue_manager.py]: If you have managed to remove all the calls to the C methods, please remove also the corresponding cwrap declarations at the beginning of the file

File: python/res/job_queue/job_queue_manager.py [hunk 6]
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
...
+
 
     def isRunning(self):
-        return self._is_running( )
+        return self.get_job_queue().is_running()
 
     def free(self):
         self._free( )
 
+    #resolve ambihuity done vs success
```
Comment (andreabrambilla) [inline python/res/job_queue/job_queue_manager.py]: typo

File: python/res/simulator/batch_simulator.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
 from ecl.util.util import BoolVector
 
+from threading import Thread
+from time import sleep
```
Comment (andreabrambilla) [inline python/res/simulator/batch_simulator.py]: why are these here?

File: python/res/job_queue/job_queue_manager.py [hunk 8]
```diff
@@ -145,16 +124,28 @@ def __repr__(self):
...
+                for job in job_queue.job_list:
+                    job.stop(job_queue.driver)
 
         for thread in started_job_threads:
             thread.join()
+
+        #Cleanup all not-started jobs
+        for job in job_queue.job_list:
+            if job.status == JobStatusType.JOB_QUEUE_WAITING:
+                job._set_status(JobStatusType.JOB_QUEUE_IS_KILLED)
```
Comment (andreabrambilla) [inline python/res/job_queue/job_queue_manager.py]: It looks like this PR includes some of the same changes in #757 
I guess you and @ManInFez should fight over it ;)

File: python/res/job_queue/job_queue_manager.py (hunk 5 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: There is a new issue for it: #728

File: python/res/job_queue/job_queue_manager.py (hunk 6 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: ok

File: python/res/simulator/batch_simulator.py (hunk 7 above)
Comment (xjules) [reply python/res/simulator/batch_simulator.py]: Legacy of testing, now it's gone ;)

File: python/res/job_queue/job_queue_manager.py (hunk 8 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: That was agreed on standup before @ManInFez went for vacation.

File: python/tests/res/simulator/test_batch_sim.py (hunk 3 above)
Comment (xjules) [reply python/tests/res/simulator/test_batch_sim.py]: ok

File: python/res/simulator/simulation_context.py (hunk 2 above)
Comment (xjules) [reply python/res/simulator/simulation_context.py]: That's for convenience purposes. Agree, this can go away.

File: python/res/job_queue/queue.py (hunk 1 above)
Comment (xjules) [reply python/res/job_queue/queue.py]: gone....

File: python/tests/res/simulator/test_simulation_context.py (hunk 4 above)
Comment (xjules) [reply python/tests/res/simulator/test_simulation_context.py]: done

File: python/tests/res/simulator/test_simulation_context.py [hunk 9]
```diff
@@ -37,30 +35,19 @@ def test_simulation_context(self):
...
-
-            with self.assertRaises(UserWarning):
-                simulation_context1.addSimulation(0, geo_id)
-
-            while simulation_context1.isRunning():
-                time.sleep(1.0)
-
-            while simulation_context2.isRunning():
-                time.sleep(1.0)
+            time.sleep(3.0)
```
Comment (markusdregi) [inline python/tests/res/simulator/test_simulation_context.py]: Please remove the `time.sleep` from all the tests and lower the `wait_until` interval to `0.1`. This should not appear in a test. If this is necessary for the code to work we should figure out why.

Comment (xjules) [reply python/tests/res/simulator/test_simulation_context.py]: ~~`time.sleep()` is unfortunately required to sync the `fs` with job states. We could perphaps integrate it into the `wait_until`~~

Comment (xjules) [reply python/tests/res/simulator/test_simulation_context.py]: This's been fixed now by waiting for the `enkf_simulation_runner.runSimpleStep` to finish.

File: python/res/job_queue/job_queue_node.py [hunk 10]
```diff
@@ -71,33 +75,40 @@ def is_running(self):
...
 
-    def run(self, driver):
+    def run(self, driver, max_submit=2):
         self.started = True
-        x = Thread(target=self.job_monitor, args=(driver, ))
+        x = Thread(target=self.job_monitor, args=(driver, max_submit))
         x.start()
         return x
 
     def stop(self,  driver):
```
Comment (oyvindeide) [inline python/res/job_queue/job_queue_node.py]: `driver` is no longer needed

File: python/res/job_queue/queue.py [hunk 11]
```diff
@@ -89,13 +90,13 @@ def __repr__(self):
         cnt = '%s, num_running=%d, num_complete=%d, num_waiting=%d, num_pending=%d, active=%d'
         return self._create_repr(cnt % (isrun, nrun, ncom, nwait, npend, len(self)))
 
-    def __init__(self, driver , max_submit=1, size=0):
+    def __init__(self, driver , max_submit=2, size=0):
         """
         Short doc...
-
+        The @max_submit argument says how many times the jon be submitted (including a failure)
```
Comment (oyvindeide) [inline python/res/job_queue/queue.py]: typo jon to job

File: python/res/job_queue/job_queue_manager.py [hunk 12]
```diff
@@ -145,16 +124,28 @@ def __repr__(self):
...
+    def max_submit(self):
+        return self.get_job_queue().get_max_submit()
+
     def execute_queue(self):
 
         job_queue = self.get_job_queue()
         started_job_threads = []
         while job_queue.is_running():
             job = job_queue.fetch_next_waiting()
             while not job_queue.stopped_by_user and job is not None and job_queue.count_running() <= self.max_running():
```
Comment (oyvindeide) [inline python/res/job_queue/job_queue_manager.py]: Doesn't this add one job more than it should?:
`job_queue.count_running() <= self.max_running()`, should've been only less than?

File: python/res/job_queue/job_queue_node.py (hunk 10 above)
Comment (ManInFez) [reply python/res/job_queue/job_queue_node.py]: driver is still used in monitoring, both for updating status and killing the job.

File: python/res/enkf/enkf_simulation_runner.py [hunk 13]
```diff
@@ -91,7 +91,6 @@ def add_job(self, run_arg, res_config, job_queue):
         num_cpu = res_config.queue_config.num_cpu
         if num_cpu == 0:
             num_cpu = res_config.ecl_config.num_cpu
-
```
Comment (markusdregi) [inline python/res/enkf/enkf_simulation_runner.py]: Please remove the white space change.

File: python/res/job_queue/job_queue_manager.py [hunk 14]
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
...
-        """
-        if job_index is None:
-            ts = self._global_progress_timestamp()
-        else:
-            ts = self._iget_progress_timestamp(job_index)
-
-        return ts.datetime()
-
-
     def get_job_queue(self):
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_manager.py]: I suggest that we in the future we rewrite this so that `queue` is a private field and then expose a property named `queue` instead of the `get_job_queue`. We could already now, to be compliant with this, use `self.queue` instead of `self.get_job_queue()` internally in job_queue_manager.

File: python/res/job_queue/job_queue_manager.py [hunk 15]
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
...
+
 
     def isRunning(self):
-        return self._is_running( )
+        return self.get_job_queue().is_running()
 
     def free(self):
         self._free( )
 
+    #resolve ambiguity done vs success
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_manager.py]: Please turn this comment into a descriptive issue

File: python/res/job_queue/job_queue_manager.py [hunk 16]
```diff
@@ -58,80 +55,62 @@ def __init__(self, queue):
...
     def isJobComplete(self, job_index):
-        return self._job_complete( job_index )
+        return not (self.get_job_queue().job_list[job_index].is_running()
+                    or self.get_job_queue().job_list[job_index].status == JobStatusType.JOB_QUEUE_WAITING)
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_manager.py]: Use `self.queue.isJobWaiting(job_index)` instead?

File: python/res/job_queue/job_queue_manager.py [hunk 17]
```diff
@@ -145,16 +124,28 @@ def __repr__(self):
     def max_running(self):
         return self.get_job_queue().get_max_running()
 
+    def max_submit(self):
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_manager.py]: Should we make this `_max_submit`?

File: python/res/job_queue/job_queue_manager.py (hunk 12 above)
Comment (markusdregi) [reply python/res/job_queue/job_queue_manager.py]: I think you are right, however this is outside the scope of the PR. Well spotted though :clap: Make an issue ;)

File: python/res/job_queue/job_queue_node.py (hunk 10 above)
Comment (markusdregi) [reply python/res/job_queue/job_queue_node.py]: @ManInFez claims that the above comment by himself can be ignored.

File: python/res/job_queue/job_queue_node.py [hunk 18]
```diff
@@ -27,14 +27,15 @@ class JobQueueNode(BaseCClass):
...
+    _get_submit_attempt = ResPrototype("int job_queue_node_get_submit_attempt(job_queue_node)")
 
     def __init__(self,job_script, job_name, run_path, num_cpu, 
                     status_file, ok_file, exit_file, 
                     done_callback_function, exit_callback_function, callback_arguments):
         self.done_callback_function = done_callback_function
         self.exit_callback_function = exit_callback_function
         self.callback_arguments = callback_arguments
-    
+        self.stopped_by_user = False
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_node.py]: Should we perhaps rename this for `stopped`? And in addition, maybe this should be a private member `_stopped`.

File: python/res/job_queue/job_queue_node.py [hunk 19]
```diff
@@ -71,33 +75,40 @@ def is_running(self):
                 self.status == JobStatusType.JOB_QUEUE_RUNNING  or
                 self.status == JobStatusType.JOB_QUEUE_UNKNOWN) # dont stop monitoring if LSF commands are unavailable
     
-    def job_monitor(self, driver):
+    def job_monitor(self, driver, max_submit):
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_node.py]: Should we make this private, i.e. `_job_monitor`?

File: python/res/job_queue/job_queue_node.py [hunk 20]
```diff
@@ -71,33 +75,40 @@ def is_running(self):
...
+            if self.submit_attempt < max_submit:
+                self._set_status(JobStatusType.JOB_QUEUE_WAITING)
+                self.started = False
+            else:
+                self._set_status(JobStatusType.JOB_QUEUE_FAILED)
+                self.run_exit_callback()
         elif self.status == JobStatusType.JOB_QUEUE_WAITING:
             self.started = False
+        elif self.status == JobStatusType.JOB_QUEUE_IS_KILLED:
+            pass
```
Comment (markusdregi) [inline python/res/job_queue/job_queue_node.py:99]: Should we add an:
```
else:
    raise AssertionError("Unexpected job status type after running job: {}".format(self.status))

File: python/res/job_queue/queue.py [hunk 21]
```diff
@@ -210,6 +211,9 @@ def get_max_job_duration(self):
     def set_max_job_duration(self, max_duration):
         self._set_max_job_duration(max_duration)
 
+    def get_max_submit(self):
```
Comment (markusdregi) [inline python/res/job_queue/queue.py]: Property?

File: python/res/simulator/batch_simulator_context.py [hunk 22]
```diff
@@ -16,7 +16,7 @@ def __init__(self, result_keys, ert, fs, mask, itr):
         """
         super(BatchContext, self).__init__(ert, fs, mask, itr)
         self.result_keys = result_keys
-        self.res_config = ert.resConfig( )
+        self.res_config = ert.resConfig()
```
Comment (markusdregi) [inline python/res/simulator/batch_simulator_context.py]: Please revert white space fixes. This PR is touching enough files as is.

File: python/res/simulator/batch_simulator_context.py [hunk 23]
```diff
@@ -78,6 +78,7 @@ def results(self):
         nodes = [ EnkfNode(self.res_config.ensemble_config[key]) for key in self.result_keys ]
         for sim_id in range(len(self)):
             node_id = NodeId( 0, sim_id)
+
```
Comment (markusdregi) [inline python/res/simulator/batch_simulator_context.py]: Please revert white space fixes. This PR is touching enough files as is.

File: python/res/simulator/simulation_context.py [hunk 24]
```diff
@@ -91,11 +79,15 @@ def didRealizationFail(self, iens):
 
 
     def isRealizationQueued(self, iens):
-        return iens in self._run_args
+        try:
```
Comment (markusdregi) [inline python/res/simulator/simulation_context.py]: Are you sure we want to do this?

File: python/res/simulator/simulation_context.py [hunk 25]
```diff
@@ -123,6 +115,8 @@ def get_run_context(self):
 
     def stop(self):
         self._queue_manager.stop_queue( )
+        if self.sim_thread is not None:
```
Comment (markusdregi) [inline python/res/simulator/simulation_context.py]: Is this necessary? Should not `self.sim_thread` always be a `Thread`?

File: python/res/simulator/batch_simulator.py [hunk 26]
```diff
@@ -204,9 +204,6 @@ def start(self, case_name, case_data):
         sim_context = BatchContext(
             self.result_keys, self.ert, file_system, mask, itr)
 
-        for sim_id, (geo_id, _) in enumerate(case_data):
```
Comment (markusdregi) [inline python/res/simulator/batch_simulator.py:207]: This scares me a bit. Could you please try to build and run Everest tests on top of this?

Comment (markusdregi) [reply python/res/simulator/batch_simulator.py:207]: In particular, the `geo_id` is lost.

File: python/res/simulator/simulation_context.py [hunk 27]
```diff
@@ -19,52 +15,44 @@ def __init__(self, ert, sim_fs, mask, itr , verbose=False):
...
-            raise UserWarning("Realization number out of range: %d >= %d" % (iens, len(self._run_context)))
-
-        if not self._mask[iens]:
-            raise UserWarning("Realization number: '%d' is not active" % iens)
-
-        if iens in self._run_args:
-            raise UserWarning("Realization number: '%d' already queued" % iens)
-
-        run_arg = self._run_context[iens]
-        run_arg.geo_id = geo_id
```
Comment (markusdregi) [inline python/res/simulator/simulation_context.py:51]: This is crucial and should be preserved. See comment on `geo_id` being lost. Could be passed to the `BatchContext` upon creation.

File: python/tests/res/job_queue/test_job_queue.py [hunk 28]
```diff
@@ -1,7 +1,9 @@
 from res.job_queue import JobStatusType, Driver, QueueDriverEnum, JobQueue, JobQueueNode
 from tests import ResTest
+from tests.utils import wait_until
```
Comment (markusdregi) [inline python/tests/res/job_queue/test_job_queue.py:3]: Because twice is always better

File: python/tests/res/job_queue/test_job_queue.py [hunk 29]
```diff
@@ -79,9 +81,13 @@ def test_kill_jobs(self):
             assert job_queue.is_running()
             
             threads = start_all(job_queue)
-         
-            job_queue.kill_all_jobs()
-            assert not job_queue.is_running()
+
```
Comment (markusdregi) [inline python/tests/res/job_queue/test_job_queue.py:83]: Should we add 

```python
time.sleep(3)
wait_until(
    lambda: self.assertTrue(job_queue.is_running())
)
```

to prevent the above `true`  vs `True` bug.

File: python/tests/res/job_queue/test_job_queue.py [hunk 30]
```diff
@@ -97,26 +103,43 @@ def test_add_jobs(self):
...
+                job.stop(job_queue.driver)
+
+            wait_until(
+                lambda: self.assertFalse(job_queue.is_running())
+            )
+
+            for t in threads:
+                t.join()
+
+            assert True
```
Comment (markusdregi) [inline python/tests/res/job_queue/test_job_queue.py]: Why?

File: python/tests/res/job_queue/test_job_queue.py [hunk 31]
```diff
@@ -97,26 +103,43 @@ def test_add_jobs(self):
...
+            job_queue = create_queue(failing_script, max_submit=1)
 
             assert job_queue.queue_size == 10
             assert job_queue.is_running()
             
             threads = start_all(job_queue)
+
+            wait_until(
+                func=(lambda: self.assertFalse(job_queue.is_running())),
+                interval=2
```
Comment (markusdregi) [inline python/tests/res/job_queue/test_job_queue.py]: Why do we need intervals here?

File: python/tests/res/job_queue/test_job_queue_manager.py [hunk 32]
```diff
@@ -24,16 +25,22 @@ def dummy_exit_callback(args):
...
 }
 
+simple_script = "#!/usr/bin/env python\n"\
+                "with open('STATUS', 'w') as f:"\
+                "   f.write('finished successfully')"\
+                "\n"
 
-def create_queue():
+failing_script = "#!/usr/bin/env python\n"\
+                        "import sys\n"\
```
Comment (markusdregi) [inline python/tests/res/job_queue/test_job_queue_manager.py]: Should we do something with the line alignment here?

Comment (markusdregi) [reply python/tests/res/job_queue/test_job_queue_manager.py]: Perhaps this would be more readable?
```python
failing_script = """
#!/usr/bin/env python
import sys
sys.exit(1)
"""
```

File: python/tests/res/simulator/test_batch_sim.py [hunk 33]
```diff
@@ -405,7 +401,11 @@ def test_stop_sim(self):
                                   })
                              ])
 
+            time.sleep(2.0)
             ctx.stop()
+            wait_until(
```
Comment (markusdregi) [inline python/tests/res/simulator/test_batch_sim.py]: Why is this necessary?

File: python/res/simulator/simulation_context.py (hunk 27 above)
Comment (markusdregi) [reply python/res/simulator/simulation_context.py:51]: Note: We should ensure that the tests actually capture this behavior

**Review by markusdregi — COMMENTED** [review]
Me and @ManInFez has gone through the entire PR now. Please address the comments.

Well done so far :+1:

File: python/res/job_queue/queue.py (hunk 11 above)
Comment (xjules) [reply python/res/job_queue/queue.py]: fixed

File: python/res/job_queue/job_queue_manager.py (hunk 12 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: yes, as was pointed out by @markusdregi it's outside this scope; mainly we count the run number from `1`

File: python/res/enkf/enkf_simulation_runner.py (hunk 13 above)
Comment (xjules) [reply python/res/enkf/enkf_simulation_runner.py]: fixed

File: python/res/job_queue/job_queue_manager.py (hunk 14 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: agree, it's fixed now:
```python
@property
def queue(self):
     return self._queue
```

File: python/res/job_queue/job_queue_manager.py (hunk 15 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: Will do ...

File: python/res/job_queue/job_queue_node.py (hunk 10 above)
Comment (xjules) [reply python/res/job_queue/job_queue_node.py]: job `stop` is fixed now.

File: python/res/job_queue/job_queue_manager.py (hunk 16 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: but then it could be still `running`, i.e., not `waiting` nor `complete` or I might have misunderstood :)

File: python/res/job_queue/job_queue_manager.py (hunk 17 above)
Comment (xjules) [reply python/res/job_queue/job_queue_manager.py]: do you mean having it as a property?
```python
@property
def _max_submit(self):
     return return self.queue.get_max_submit()
```

File: python/res/job_queue/job_queue_node.py (hunk 18 above)
Comment (xjules) [reply python/res/job_queue/job_queue_node.py]: I think it's just a legacy from the past. Agree, this could be renamed.

Comment (xjules) [reply python/res/job_queue/job_queue_node.py]: fixed now

File: python/res/job_queue/job_queue_node.py (hunk 19 above)
Comment (xjules) [reply python/res/job_queue/job_queue_node.py]: sure

File: python/res/job_queue/job_queue_node.py (hunk 20 above)
Comment (xjules) [reply python/res/job_queue/job_queue_node.py:99]: We can add this, but we should also clear out the job states we want to use

File: python/res/job_queue/queue.py (hunk 21 above)
Comment (xjules) [reply python/res/job_queue/queue.py]: `max_submit` property now resides in queue only and `queue_manager` pokes queue for it, so `max_submit` function is gone from `queue_manager`

File: python/res/simulator/batch_simulator_context.py (hunk 22 above)
Comment (xjules) [reply python/res/simulator/batch_simulator_context.py]: ok ... just didn't like the emptiness inside :)

File: python/res/simulator/batch_simulator_context.py (hunk 23 above)
Comment (xjules) [reply python/res/simulator/batch_simulator_context.py]: fixed

File: python/res/simulator/simulation_context.py (hunk 24 above)
Comment (xjules) [reply python/res/simulator/simulation_context.py]: we decided not have self._run_args any more (since it's in run_context) - will try to beautify a bit.

File: python/res/simulator/simulation_context.py (hunk 25 above)
Comment (xjules) [reply python/res/simulator/simulation_context.py]: ok, that was the legacy from the previous version where `self.sim_thread` was initialized independently. fixed now

File: python/res/simulator/batch_simulator.py (hunk 26 above)
Comment (xjules) [reply python/res/simulator/batch_simulator.py:207]: now `case_data` is an extra argument to `simulation_context`

File: python/res/simulator/simulation_context.py (hunk 27 above)
Comment (xjules) [reply python/res/simulator/simulation_context.py:51]: Agree, we can sit down and suggest a proper test for capturing the `geo_id`

File: python/tests/res/job_queue/test_job_queue.py (hunk 28 above)
Comment (xjules) [reply python/tests/res/job_queue/test_job_queue.py:3]: agree, twice is always better

File: python/tests/res/job_queue/test_job_queue_manager.py (hunk 32 above)
Comment (xjules) [reply python/tests/res/job_queue/test_job_queue_manager.py]: readability fixed

File: python/tests/res/job_queue/test_job_queue.py (hunk 30 above)
Comment (xjules) [reply python/tests/res/job_queue/test_job_queue.py]: fixed now

File: python/tests/res/job_queue/test_job_queue.py (hunk 31 above)
Comment (xjules) [reply python/tests/res/job_queue/test_job_queue.py]: Could be removed I reckon.

File: python/tests/res/simulator/test_batch_sim.py (hunk 33 above)
Comment (xjules) [reply python/tests/res/simulator/test_batch_sim.py]: `context.stop()` is blocking operation now, but needs to be tested a few times first - then it's gone

File: python/tests/res/job_queue/test_job_queue.py (hunk 29 above)
Comment (xjules) [reply python/tests/res/job_queue/test_job_queue.py:83]: fixed

File: python/res/simulator/simulation_context.py (hunk 27 above)
Comment (xjules) [reply python/res/simulator/simulation_context.py:51]: Decided just to do a simple `geo_id` assert in `test_simulation_context`

**Review by ManInFez — APPROVED** [review]
A very big step in removing C from the queue system. Good work!

File: python/tests/res/simulator/test_simulation_context.py [hunk 34]
```diff
@@ -29,38 +27,28 @@ def test_simulation_context(self):
...
             first_half = fs_manager.getFileSystem("first_half")
             other_half = fs_manager.getFileSystem("other_half")
 
-            simulation_context1 = SimulationContext(ert, first_half, mask1 , 0)
-            simulation_context2 = SimulationContext(ert, other_half, mask2 , 0)
-
-            ert.createRunpath( simulation_context1.get_run_context( ) )
-            ert.createRunpath( simulation_context2.get_run_context( ) )
+            # i represents geo_id
+            case_data = [(i,0) for i in range(size)]
```
Comment (markusdregi) [inline python/tests/res/simulator/test_simulation_context.py]: This is wrong usage of the API. Please revise the documentation: https://github.com/equinor/libres/blob/9ba9839785764568e6479b34005154bfb8b7575f/python/res/simulator/batch_simulator.py#L133

The following should be correct:
```python
case_data = [(i, {}) for i in range(size)]
```

Comment (xjules) [reply python/tests/res/simulator/test_simulation_context.py]: ok, I knew it was supposed to be a dict, but since it not used anyway ... fixed now

**Comment by ManInFez** [comment]
test this please!
